### PR TITLE
#1723 - E2E/Units Tests - Review Workers E2E Tests

### PR DESCRIPTION
**Pull request by andrepestana-aot**
- Added a createTestingAppModule for Workers E2E tests;
- Moved disbursement.controller.saveDisbursementSchedules.e2e-spec to e2e folder;
- Create the following tests:
	- apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts
	ProgramInfoRequestController(e2e)-updateApplicationStatus
		✓ Should update application PIR status when not set. (67 ms)
		✓ Should not update application PIR status when already set. (22 ms)
		✓ Should have job error status when could not find the application. (2 ms)
	
	- apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts
	SupportingUserController(e2e)-createSupportingUsers
		✓ Should create parent supporting users when requested. (87 ms)
		✓ Should create partner supporting user when requested. (39 ms)
		✓ Should not create supporting users when application already has one. (45 ms)
	
	- apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts
	ApplicationController(e2e)-updateStatus
		✓ Should update application status when it receives the correct original status. (80 ms)
		✓ Should not update application status and return error when it does not receive the correct original status. (35 ms)
		✓ Should return error when it does not find the application. (2 ms)
	
	- apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts
	ApplicationController(e2e)-verifyApplicationExceptions
		✓ Should create and associate application exceptions when there are some exceptions for the application. (57 ms)
		✓ Should not create any application exception when there is no application exception in application data. (13 ms)
		✓ Should return error when it does not find the application. (3 ms)
	
	- apps/workers/test/app.e2e-spec.ts
	WorkersController (e2e)
		✓ Should be able to initialize workers module (564 ms)

### CODE REVIEW

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+} from "../../application.dto";
+import { createFakeUpdateApplicationStatusPayload } from "./update-application-status";
+
+describe("ApplicationController(e2e)-updateStatus", () => {
+  let db: E2EDataSources;
+  let applicationController: ApplicationController;
+
+  beforeAll(async () => {
+    const { nestApplication, dataSource } = await createTestingAppModule();
+    await nestApplication.init();
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts:30]: I believe that the `nestApplication.init()` is already happening in `createTestingAppModule`.

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      >(updateApplicationStatusPayload),
+    );
+
+    // Asserts
+    expect(result).toHaveProperty(
+      FAKE_WORKER_JOB_RESULT_PROPERTY,
+      MockedZeebeJobResult.Complete,
+    );
+
+    // Asserts that the application status has changed to declined
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts:61]: Please add the period.

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    const expectedApplication = await db.application.findOneBy({
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts:62]: As a suggestion I would prefer to load only what we would be testing.
```ts
const expectedApplication = await db.application.findOne({
  select: { applicationStatus: true },
  where: { id: savedApplication.id },
});
```

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      id: savedApplication.id,
+    });
+    expect(expectedApplication.applicationStatus).toEqual(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts:65]: Did the `toBe` not work for this case?

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    );
+
+    // Asserts
+    expect(result).toHaveProperty(
+      FAKE_WORKER_JOB_RESULT_PROPERTY,
+      MockedZeebeJobResult.Complete,
+    );
+
+    const updatedApplication = await db.application.findOne({
+      relations: { applicationException: true },
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:64]: Can we add the select here also?

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      where: {
+        id: savedApplication.id,
+      },
+    });
+    expect(updatedApplication.applicationException).toHaveProperty(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:69]: Is there a reason to not use the `expect` as below?
```ts
expect(updatedApplication.applicationException.exceptionStatus).toBe(
  ApplicationExceptionStatus.Pending,
);
```

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:1]: Maybe not part of the PR since the main goal was to start the E2E for workers but the scenario where "The exceptions were already processed for this application." is important to ensure the idempotency of the process.
We can add it to this PR os create a new ticket to capture the next ones.

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    // Asserts
+    expect(result).toEqual({
+      [FAKE_WORKER_JOB_RESULT_PROPERTY]: MockedZeebeJobResult.Error,
+      [FAKE_WORKER_JOB_ERROR_MESSAGE_PROPERTY]:
+        "The application status was not updated either because the application id was not found or the application is not in the expected status.",
+      [FAKE_WORKER_JOB_ERROR_CODE_PROPERTY]: APPLICATION_STATUS_NOT_UPDATED,
+    });
+
+    // Asserts that the application status has not changed to declined
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-updateStatus.e2e-spec.ts:98]: https://github.com/bcgov/SIMS/pull/1941/files#r1199406564

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/update-application-status.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  const variables = {
+    applicationId,
+  };
+  return createFakeWorkerJob<
+    ApplicationUpdateStatusJobInDTO,
+    ApplicationUpdateStatusJobHeaderDTO,
+    IOutputVariables
+  >({
+    variables,
+    customHeaders: { fromStatus, toStatus },
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/update-application-status.ts:36]: Since `customHeaders` is following this format I would prefer to have the variables doing the same.
```ts
variables: { applicationId },
customHeaders: { fromStatus, toStatus },
```

File: sources/packages/backend/apps/workers/src/controllers/disbursement/_tests_/e2e/disbursement.controller.saveDisbursementSchedules.e2e-spec.ts [hunk 9]
```diff
@@ -1,78 +1,50 @@
...
+import { DisbursementController } from "../../disbursement.controller";
 import {
   FAKE_WORKER_JOB_RESULT_PROPERTY,
   MockedZeebeJobResult,
-} from "../../../../test/utils/worker-job-mock";
+} from "../../../../../test/utils/worker-job-mock";
+import { createTestingAppModule } from "../../../../testHelpers";
 
-describe("Disbursement Schedule Service - Create disbursement", () => {
-  let userRepo: Repository<User>;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/disbursement/_tests_/e2e/disbursement.controller.saveDisbursementSchedules.e2e-spec.ts:37]: Really great simplification. This one serves as an example of how the use of the `E2EDataSources` can make the E2E code cleaner 😉

File: sources/packages/backend/apps/workers/src/controllers/disbursement/_tests_/e2e/disbursement.controller.saveDisbursementSchedules.e2e-spec.ts [hunk 10]
```diff
@@ -192,24 +164,36 @@ describe("Disbursement Schedule Service - Create disbursement", () => {
...
+        valueAmount: 350,
+        disbursedAmountSubtracted: 350,
+      },
+    );
     // Grants do not generate overaward. The the Canada grant CSGP got 1500 already
     // disbursed value so the student already received $300 extra. For this case the
     // student will not receive the money for the grant but will no be charged either
     // for the $300 already received.
-    assertAwardDeduction(
+    await assertAwardDeduction(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/disbursement/_tests_/e2e/disbursement.controller.saveDisbursementSchedules.e2e-spec.ts:196]: Nice catch 👍

File: sources/packages/backend/apps/workers/src/testHelpers/testing-modules/testing-modules-helper.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/testHelpers/testing-modules/testing-modules-helper.ts:1]: Can we please follow the folder structure from the `queue-consumers`?
I believe that this is following the API one that is old and needs to be changed. For instance, we have not been using camel cases (e.g. testHelpers) for folder names on the backend lately.

File: sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Creates a fake supporting user.
+ * @param supportingUserType supporting user type to be created.
+ * @param relations dependencies:
+ * - `application`: application that the supporting user is associated with.
+ * @returns a fake supporting user.
+ */
+export function createFakeSupportingUser(
+  supportingUserType: SupportingUserType,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts:12]: Providing `supportingUserType` here would be the same case that we discussed in the past.
If there is a Team desire we can revisit the topic.

File: sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Saves a fake supporting user.
+ * @param dataSource data source for the application.
+ * @param supportingUserType supporting user type to be created.
+ * @param relations dependencies:
+ * - `application`: application that the supporting user is associated with.
+ * @returns a persisted fake supporting user.
+ */
+export async function saveFakeSupportingUser(
+  dataSource: DataSource,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts:34]: For the new methods we can use the `E2EDataSources`, IMO.

File: sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Saves a fake supporting user.
+ * @param dataSource data source for the application.
+ * @param supportingUserType supporting user type to be created.
+ * @param relations dependencies:
+ * - `application`: application that the supporting user is associated with.
+ * @returns a persisted fake supporting user.
+ */
+export async function saveFakeSupportingUser(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts:33]: The idea of creating the `saveFakeSomething` method was to deal with a situation where there is a proper sequence to save dependent entities to the DB.
Usually, we would not create these types of methods when the only intention is to call the `createFakeSomething` and have it saved.

File: sources/packages/backend/apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+      >(updateApplicationStatusPayload),
+    );
+
+    // Asserts
+    expect(result).toHaveProperty(
+      FAKE_WORKER_JOB_RESULT_PROPERTY,
+      MockedZeebeJobResult.Complete,
+    );
+
+    // Asserts that the application PIR status has changed to declined
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts:72]: Period.

File: sources/packages/backend/apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    );
+
+    const updatedApplication = await db.application.findOne({
+      relations: { supportingUsers: true },
+      where: {
+        id: savedApplication.id,
+      },
+    });
+    expect(updatedApplication.supportingUsers).toHaveLength(2);
+    expect(updatedApplication.supportingUsers[0]).toHaveProperty(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts:63]: It can also be referenced as below:
```ts
const [parent1, parent2] = updatedApplication.supportingUsers;
```

File: sources/packages/backend/apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+  let db: E2EDataSources;
+  let supportingUserController: SupportingUserController;
+
+  beforeAll(async () => {
+    const { nestApplication, dataSource } = await createTestingAppModule();
+    db = createE2EDataSources(dataSource);
+    supportingUserController = nestApplication.get(SupportingUserController);
+  });
+
+  it("Should create parent supporting users when requested.", async () => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts:32]: I would change the descriptions to indicate multiple parents: "Should create parents supporting users when requested.".

**Review by andrewsignori-aot — COMMENTED** [review]
Great work and thanks for extending the work across multiple controllers. Please take a look at the comments.

**Review by andrewsignori-aot — COMMENTED** [review]
Great work and thanks for extending the work across multiple controllers. Please take a look at the comments.

**Review by ann-aot — COMMENTED** [review]
Good work @andrepestana-aot. Mostly similar comments as @andrewsignori-aot

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:64]: Not sure if it's my GIT issue. I am seeing the changes

File: sources/packages/backend/apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    );
+
+    // Asserts
+    expect(result).toHaveProperty(
+      FAKE_WORKER_JOB_RESULT_PROPERTY,
+      MockedZeebeJobResult.Complete,
+    );
+
+    // Asserts that the application PIR status has not changed
+    const expectedApplication = await db.application.findOneBy({
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts:115]: https://github.com/bcgov/SIMS/pull/1941/files#r1199412241

File: sources/packages/backend/apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+      >(updateApplicationStatusPayload),
+    );
+
+    // Asserts
+    expect(result).toHaveProperty(
+      FAKE_WORKER_JOB_RESULT_PROPERTY,
+      MockedZeebeJobResult.Complete,
+    );
+
+    // Asserts that the application PIR status has not changed
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/controllers/program-info-request/_tests_/e2e/program-info-request.controller-updateApplicationStatus.e2e-spec.ts:114]: period

File: sources/packages/backend/apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+    );
+
+    const updatedApplication = await db.application.findOne({
+      relations: { supportingUsers: true },
+      where: {
+        id: savedApplication.id,
+      },
+    });
+    expect(updatedApplication.supportingUsers).toHaveLength(1);
+    expect(updatedApplication.supportingUsers[0]).toHaveProperty(
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/controllers/supporting-user/_test_/e2e/supporting-user.controller.createSupportingUsers.e2e-spec.ts:104]: if we follow line 63 logic, we can avoid indexing. same for 147, IMO

File: sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts [hunk 20]
```diff
@@ -0,0 +1,19 @@
+import { Application, SupportingUser, SupportingUserType } from "@sims/sims-db";
+
+/**
+ * Creates a fake supporting user.
+ * @param supportingUserType supporting user type to be created.
```
Comment (ann-aot) [inline sources/packages/backend/libs/test-utils/src/factories/supporting-user.ts:5]: pls review the comment

**Review by ann-aot — COMMENTED** [review]
some minor comments added

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:64]: I am not able to see any change here also. Was the code changed?

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes. Please take a look at the last minor comments.

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts
Comment (andrepestana-aot) [reply sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:1]: I created a test for this case: "Should not create an application exception when there is already one for the application."

File: sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts (hunk 5 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/workers/src/controllers/application/_tests_/e2e/application.controller-verifyApplicationExceptions.e2e-spec.ts:65]: I added a "select" for all queries now. Thanks.

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes 👍  Good work @andrepestana-aot

**Review by sh16011993 — APPROVED** [review]
LGTM 👍

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, looks good 👍
